File: ts/src/engine/gates.ts
/**
 * Mechanical (non-LLM) gate framework, ported from the Python engine.
 *
 * A gate is a named deterministic check over the primary document text.
 * Gates never consult a model; failures become results, never crashes.
 * Finding strings are parity-pinned against fixtures/parity/gate_cases.json.
 */

import { pyRepr } from "./pyformat.js";

export interface GateResult {
  passed: boolean;
  findings: string[];
}

export type GateFn = (primaryText: string, project: unknown) => GateResult;

export interface Gate {
  name: string;
  fn: GateFn;
}

export function runGate(gate: Gate, primaryText: string, project: unknown = null): GateResult {
  try {
    return gate.fn(primaryText, project);
  } catch (exc) {
    return {
      passed: false,
      findings: [`gate '${gate.name}' crashed: ${exc instanceof Error ? exc.message : exc}`],
    };
  }
}

/** Run every gate; failures become results, never crashes. */
export function runGates(
  gates: Gate[],
  primaryText: string,
  project: unknown = null,
): Record<string, GateResult> {
  const results: Record<string, GateResult> = {};
  for (const gate of gates) {
    results[gate.name] = runGate(gate, primaryText, project);
  }
  return results;
}

export function allGatesPassed(results: Record<string, GateResult>): boolean {
  return Object.values(results).every((r) => r.passed);
}

// ---------------------------------------------------------------------------
// Gate batteries
// ---------------------------------------------------------------------------

const WORD_RE = /\S+/g;

function countWords(text: string): number {
  return (text.match(WORD_RE) ?? []).length;
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/** Python str.splitlines() line-break set (close enough for document text). */
function splitLines(text: string): string[] {
  if (text === "") return [];
  const lines = text.split(/\r\n|[\n\r\v\f\x1c\x1d\x1e\x85\u2028\u2029]/);
  // splitlines() drops a trailing empty segment after a final line break.
  if (lines.length > 0 && lines[lines.length - 1] === "") lines.pop();
  return lines;
}

/** Extract a markdown section's body by heading title (any # level). */
function extractSection(text: string, section: string): string | null {
  const headingRe = new RegExp(`^(#{1,6})\\s+${escapeRegExp(section)}\\s*$`, "im");
  const m = headingRe.exec(text);
  if (m === null) return null;
  const level = m[1]!.length;
  const rest = text.slice(m.index + m[0].length);
  const nextRe = new RegExp(`^#{1,${level}}\\s+\\S`, "m");
  const nxt = nextRe.exec(rest);
  return nxt ? rest.slice(0, nxt.index) : rest;
}

/**
 * Fail when the counted region exceeds maxWords. Counting scope (mutually
 * exclusive): whole document (default), one named markdown `section`, or
 * only the remainder of lines starting with `linePrefix` (e.g. "VO:").
 */
export function wordCountGate(
  maxWords: number,
  opts: { section?: string | null; linePrefix?: string | null } = {},
): Gate {
  const section = opts.section ?? null;
  const linePrefix = opts.linePrefix ?? null;
  if (section !== null && linePrefix !== null) {
    throw new Error("wordCountGate: pass section OR linePrefix, not both");
  }

  const name = section
    ? `word_count_${section.toLowerCase().replace(/ /g, "_")}`
    : "word_count";
  const prefixLabel = linePrefix ? linePrefix.replace(/:+$/, "").trim() : null;

  const fn: GateFn = (primaryText) => {
    let target = primaryText;
    let scope = "document";
    if (section !== null) {
      const extracted = extractSection(primaryText, section);
      if (extracted === null) {
        return {
          passed: false,
          findings: [
            `section '${section}' not found — cannot verify its word count ` +
              `(missing section fails the gate)`,
          ],
        };
      }
      target = extracted;
      scope = `section '${section}'`;
    } else if (linePrefix !== null) {
      const matchedLines = splitLines(primaryText)
        .map((line) => line.trimStart())
        .filter((line) => line.startsWith(linePrefix))
        .map((line) => line.slice(linePrefix.length));
      if (matchedLines.length === 0) {
        return {
          passed: false,
          findings: [
            `no lines start with '${linePrefix}' — cannot verify the ` +
              `${prefixLabel} word count (missing content fails the gate)`,
          ],
        };
      }
      const count = countWords(matchedLines.join("\n"));
      if (count > maxWords) {
        return {
          passed: false,
          findings: [`${count} ${prefixLabel} words > ${maxWords} (max ${maxWords})`],
        };
      }
      return { passed: true, findings: [] };
    }
    const count = countWords(target);
    if (count > maxWords) {
      return {
        passed: false,
        findings: [`${scope} is ${count} words (max ${maxWords})`],
      };
    }
    return { passed: true, findings: [] };
  };

  return { name, fn };
}

/**
 * Fail on known-bad aliases of canonical terms. Alias matching is
 * case-sensitive and word-boundary based ((?<!\w)alias(?!\w)).
 *
 * Parity note: Python's \w is Unicode; JS's is ASCII. Identical for ASCII
 * terms — keep non-ASCII canonical terms in mind if they ever appear.
 */
export function termLintGate(canonicalTerms: Record<string, string[]>): Gate {
  const fn: GateFn = (primaryText) => {
    const findings: string[] = [];
    for (const [canonical, aliases] of Object.entries(canonicalTerms)) {
      for (const alias of aliases) {
        if (alias === canonical) continue;
        const pattern = new RegExp(`(?<!\\w)${escapeRegExp(alias)}(?!\\w)`, "g");
        const hits = primaryText.match(pattern) ?? [];
        if (hits.length > 0) {
          findings.push(
            `found '${alias}' ×${hits.length} — canonical spelling is '${canonical}'`,
          );
        }
      }
    }
    return { passed: findings.length === 0, findings };
  };
  return { name: "term_lint", fn };
}

/** Fail when any banned regex pattern appears in the document. */
export function bannedElementsGate(patterns: string[]): Gate {
  const compiled = patterns.map((p) => new RegExp(p, "m"));
  const fn: GateFn = (primaryText) => {
    const findings: string[] = [];
    for (let i = 0; i < patterns.length; i++) {
      const m = compiled[i]!.exec(primaryText);
      if (m) {
        const snippet = splitLines(primaryText.slice(m.index, m.index + 60))[0] ?? "";
        findings.push(
          `banned pattern ${pyRepr(patterns[i]!)} matched at char ${m.index}: ` +
            `${pyRepr(snippet)}`,
        );
      }
    }
    return { passed: findings.length === 0, findings };
  };
  return { name: "banned_elements", fn };
}

File: src/quorable/engine/gates.py
"""Mechanical (non-LLM) gate framework.

A gate is `Gate(name, fn)` where `fn(primary_text, project) -> GateResult`.
Gates never consult a model — they exist to catch the class of defect that is
cheap to detect deterministically (overlength scripts, banned terms, term
drift), exactly like the parent's citation/xref gates did for filings.

The engine provides combinators plus three batteries packs can instantiate:
`word_count_gate`, `term_lint_gate`, `banned_elements_gate`. Packs are free
to add their own Gate objects with arbitrary callables.
"""
from __future__ import annotations

import logging
import re
from dataclasses import dataclass, field
from typing import Any, Callable

logger = logging.getLogger(__name__)


@dataclass
class GateResult:
    """The outcome of running one mechanical gate."""

    passed: bool
    findings: list[str] = field(default_factory=list)


# fn(primary_text, project) -> GateResult. `project` is an opaque handle
# (usually the Config, may be None) so pack gates can reach project data
# without the engine dictating its shape.
GateFn = Callable[[str, Any], GateResult]


@dataclass
class Gate:
    """A named mechanical check over the primary document text."""

    name: str
    fn: GateFn

    def run(self, primary_text: str, project: Any = None) -> GateResult:
        try:
            return self.fn(primary_text, project)
        except Exception as exc:  # noqa: BLE001 — a broken gate is a failed gate
            logger.error("Gate '%s' raised: %s", self.name, exc)
            return GateResult(
                passed=False,
                findings=[f"gate '{self.name}' crashed: {exc}"],
            )


def run_gates(
    gates: list[Gate],
    primary_text: str,
    project: Any = None,
) -> dict[str, GateResult]:
    """Run every gate; failures become results, never crashes."""
    results: dict[str, GateResult] = {}
    for gate in gates:
        result = gate.run(primary_text, project)
        results[gate.name] = result
        if not result.passed:
            logger.warning(
                "Gate '%s' FAILED: %s", gate.name, "; ".join(result.findings),
            )
    return results


def all_gates_passed(results: dict[str, GateResult]) -> bool:
    return all(r.passed for r in results.values())


# ---------------------------------------------------------------------------
# Gate batteries
# ---------------------------------------------------------------------------

_WORD_RE = re.compile(r"\S+")


def _extract_section(text: str, section: str) -> str | None:
    """Extract a markdown section's body by heading title (any # level).

    Returns None when the heading is absent. The section runs until the next
    heading of the same or shallower level.
    """
    heading_re = re.compile(
        rf"^(#{{1,6}})\s+{re.escape(section)}\s*$", re.MULTILINE | re.IGNORECASE,
    )
    m = heading_re.search(text)
    if m is None:
        return None
    level = len(m.group(1))
    rest = text[m.end():]
    next_re = re.compile(rf"^#{{1,{level}}}\s+\S", re.MULTILINE)
    nxt = next_re.search(rest)
    return rest[: nxt.start()] if nxt else rest


def word_count_gate(
    max_words: int,
    section: str | None = None,
    line_prefix: str | None = None,
) -> Gate:
    """Fail when the counted region exceeds max_words.

    Counting scope (mutually exclusive):
    - default: the whole document,
    - `section`: one named markdown section,
    - `line_prefix`: only the remainder of lines starting with the prefix
      (e.g. `line_prefix="VO:"` counts voice-over words in a script whose
      other lines are shot descriptions and overlays).

    Gate name is "word_count" (or "word_count_<section>") so golden-manifest
    detector keys stay stable when the limit is tuned.
    """
    if section is not None and line_prefix is not None:
        raise ValueError("word_count_gate: pass section OR line_prefix, not both")

    name = (
        f"word_count_{section.lower().replace(' ', '_')}"
        if section
        else "word_count"
    )
    prefix_label = line_prefix.rstrip(":").strip() if line_prefix else None

    def _fn(primary_text: str, project: Any) -> GateResult:
        target = primary_text
        scope = "document"
        if section is not None:
            extracted = _extract_section(primary_text, section)
            if extracted is None:
                return GateResult(
                    passed=False,
                    findings=[
                        f"section '{section}' not found — cannot verify its "
                        f"word count (missing section fails the gate)"
                    ],
                )
            target = extracted
            scope = f"section '{section}'"
        elif line_prefix is not None:
            matched_lines = [
                line.lstrip()[len(line_prefix):]
                for line in primary_text.splitlines()
                if line.lstrip().startswith(line_prefix)
            ]
            if not matched_lines:
                return GateResult(
                    passed=False,
                    findings=[
                        f"no lines start with '{line_prefix}' — cannot verify "
                        f"the {prefix_label} word count (missing content "
                        f"fails the gate)"
                    ],
                )
            target = "\n".join(matched_lines)
            count = len(_WORD_RE.findall(target))
            if count > max_words:
                return GateResult(
                    passed=False,
                    findings=[
                        f"{count} {prefix_label} words > {max_words} "
                        f"(max {max_words})"
                    ],
                )
            return GateResult(passed=True)
        count = len(_WORD_RE.findall(target))
        if count > max_words:
            return GateResult(
                passed=False,
                findings=[f"{scope} is {count} words (max {max_words})"],
            )
        return GateResult(passed=True)

    return Gate(name=name, fn=_fn)


def term_lint_gate(canonical_terms: dict[str, list[str]]) -> Gate:
    """Fail on known-bad aliases of canonical terms.

    `canonical_terms` maps the canonical spelling to a list of known-bad
    aliases. Alias matching is case-sensitive and word-boundary based, so a
    lowercase alias of a capitalized canonical term is caught without
    flagging the canonical spelling itself.
    """

    def _fn(primary_text: str, project: Any) -> GateResult:
        findings: list[str] = []
        for canonical, aliases in canonical_terms.items():
            for alias in aliases:
                if alias == canonical:
                    continue
                pattern = re.compile(rf"(?<!\w){re.escape(alias)}(?!\w)")
                hits = pattern.findall(primary_text)
                if hits:
                    findings.append(
                        f"found '{alias}' ×{len(hits)} — canonical spelling "
                        f"is '{canonical}'"
                    )
        return GateResult(passed=not findings, findings=findings)

    return Gate(name="term_lint", fn=_fn)


def banned_elements_gate(patterns: list[str]) -> Gate:
    """Fail when any banned regex pattern appears in the document."""
    compiled = [re.compile(p, re.MULTILINE) for p in patterns]

    def _fn(primary_text: str, project: Any) -> GateResult:
        findings: list[str] = []
        for raw, pattern in zip(patterns, compiled):
            m = pattern.search(primary_text)
            if m:
                snippet = primary_text[m.start(): m.start() + 60].splitlines()[0]
                findings.append(
                    f"banned pattern {raw!r} matched at char {m.start()}: "
                    f"{snippet!r}"
                )
        return GateResult(passed=not findings, findings=findings)

    return Gate(name="banned_elements", fn=_fn)
